### PR TITLE
os/chain_xattr: cope with xfs limit of 254 bytes for inlined xattrs.

### PR DESCRIPTION
**Pull request by liewegas**

### CODE REVIEW

**Comment by liewegas** [comment]
@mslovy how does this look?  (untested, and really needs some tests in test/objectstore/chain_xattr.cc, if you have the time!)

**Comment by liewegas** [comment]
added tests, passes make check

**Comment by mslovy** [comment]
@liewegas reasonable,  I agree that it becomes awkward if it breaks into too many 250bytes attrs. Xfs does linear search if the number of xattr items is small. But too many xattr items will result in a BTree Extent Struct (which is also not inline) in order to improve the search speed. I remember the critical point it changes to a BTree Format is 10, but not sure now. So We may set CHAIN_XATTR_SHORT_LEN_THRESHOLD this value smaller (say, 1000 or 750)?

**Comment by yuyuyu101** [comment]
I wonder whether 254 bytes or 10 items spill out value are configurated or hard-coded?

It looks a little inflexible to let hardcoded a spill out value in ceph? maybe we can detect it via xfs command?

**Comment by liewegas** [comment]
changed it to 1000 (4x 250byte stripes) and repushed.

i don't think it's a good idea to make this tunable, since the read path explicitly looks for multiples of the stripe length and we will fail to read old ones if the values change.  it is annoying that this is working around an XFS thing, though.  :/  but, that's what almost everyone is using with FileStore..

**Comment by yuyuyu101** [comment]
Oh, I think my meaning is whether 254 bytes and 10 items can be configurated in xfs side. So if we can improve 254 to 512 bytes, it should be enough to store ceph internal xattr in one call.

**Comment by liewegas** [comment]
oh.  in xfs the limit is fixed because of the on-disk format.  they are 
using a single byte to encode the length of the xattr value.  the inode 
size can change, which effects how small attrs we can store...  right now 
we recommend (and the tools use) 2k.

File: src/os/chain_xattr.cc [hunk 1]
```diff
@@ -135,25 +136,32 @@ int chain_getxattr(const char *fn, const char *name, void *val, size_t size)
   do {
     chunk_size = (size < CHAIN_XATTR_MAX_BLOCK_LEN ? size : CHAIN_XATTR_MAX_BLOCK_LEN);
     get_raw_xattr_name(name, i, raw_name, sizeof(raw_name));
-    size -= chunk_size;
 
     r = sys_getxattr(fn, raw_name, (char *)val + pos, chunk_size);
+    if (i && r == -ENODATA) {
```
Comment (yuyuyu101) [inline src/os/chain_xattr.cc]: ENODATA may be ENOATTR better

BSD, Linux, DARWIN all with return ENOATTR and ENOATTR is defined to be ENODATA  in Linux

**Comment by liewegas** [comment]
@athanatos I think we should run this through qa?

**Comment by dachary** [comment]
please ignore the last bot failure (network problem)

**Comment by athanatos** [comment]
error: os/chain_xattr.cc:141:20: 'ENOATTR' was not declared in this scope

**Comment by yuyuyu101** [comment]
@liewegas It seemed that we need to wrap ENOATTR or ENODATA at lower level:

1) return ENOATTR, while ENODATA is defined and has other usages: NetBSD
and MacOS X

2) return ENODATA, while ENOATTR is not defined: Linux

3) return ENOATTR, while ENODATA is not defined: FreeBSD

Linux: (ENOATTR is defined to be a synonym for ENODATA in <attr/xattr.h>

Since only linux doesn't define ENOATTR and <attr/xattr.h> is included in attr package, maybe we define ENOATTR to ENODATA in common/xattr.c for linux?

**Comment by liewegas** [comment]
Can we just s/ENODATA/ENOATTR/ everywhere then?  

> 2) return ENODATA, while ENOATTR is not defined: Linux

Linux does #define ENOATTR ENODATA, right?

**Comment by liewegas** [comment]
> error: os/chain_xattr.cc:141:20: 'ENOATTR' was not declared in this scope

Oh, i see.

Yeah, let's do

#ifndef ENOATTR
# define ENOATTR ENODATA
#endif

in our errno.h?

**Comment by yuyuyu101** [comment]
Yeah, it looks ok.

**Comment by liewegas** [comment]
added the define and rebased

**Pull request merged**